lib/rubygem_fs.rb

```diff
@@ -57,33 +64,49 @@ def each_key(prefix: "", &)
     end
 
     def remove(*keys)
-      keys.flatten.reject { |key| delete(key) }
+      keys.flatten.reject do |key|
+        delete metadata_key(key)
+        delete key
+      end
     end
 
     private
 
     def delete(key)
-      path = path_for(key)
-      return false unless descendant?(path)
-      path.ascend do |entry|
-        break unless descendant?(entry)
-        entry.delete
-      end
+      path_for(key).ascend { |entry| descendant?(entry) ? entry.delete : break }
       true
     rescue Errno::ENOTEMPTY
       true
     rescue Errno::ENOENT
       false
     end
 
-    def path_for(key)
+    def path_for(key, base = @base_dir)
       key = key.sub(%r{^/+}, "")
-      @base_dir.join(key).expand_path
+      path = base.join(key).expand_path
+      raise UnsafePathError, "Unsafe path: #{key} is outside of base: #{@base_dir}" unless descendant?(path)
+      path
     end
 
     def descendant?(path)
       path.to_s.start_with?(@base_dir.to_s) && path != @base_dir
     end
+
+    def metadata_key(key)
+      "_metadata/#{key}.json"
+    end
+
+    def metadata_write(key, **metadata)
+      path = path_for metadata_key(key)
+      path.dirname.mkpath
+      path.write JSON.generate(metadata.merge(key: key))
+    end
+
+    def metadata_read(key)
+      JSON.parse(path_for(metadata_key(key)).read).symbolize_keys
+    rescue Errno::ENOENT, JSON::ParserError
+      { key: key, metadata: {} }
```

Comment:
should this also bubble the error up? or `Rails.error.report` it?

Reply:
This only happens in tests. It would be normal for metadata to be missing if none was saved (like when gems are saved). I can't imagine the parse error actually happening but it seemed harmless.